includes/compat/class-wc-stripe-subscriptions-helper.php

```diff
@@ -63,16 +85,66 @@ public static function get_some_detached_subscriptions() {
 						'customer_id'               => $subscription->get_meta( '_stripe_customer_id' ),
 						'change_payment_method_url' => $subscription->get_change_payment_method_url(),
 					];
-					if ( count( $detached_subscriptions ) >= 5 ) {
-						break;
-					}
 				}
 			}
 		}
 
 		// Cache the result for a day.
-		set_transient( self::DETACHED_SUBSCRIPTIONS_TRANSIENT_KEY, $detached_subscriptions, DAY_IN_SECONDS );
+		WC_Stripe_Database_Cache::set( self::DETACHED_SUBSCRIPTIONS_CACHE_KEY, $detached_subscriptions, DAY_IN_SECONDS );
 
 		return $detached_subscriptions;
 	}
+
+	/**
+	 * Builds a string containing messages about subscriptions that are detached from the customer.
+	 *
+	 * @param $subscriptions array An array of subscriptions that are detached from the customer.
+	 * @return string A string containing the messages to be displayed in the admin interface.
+	 */
+	public static function build_subscriptions_detached_messages( $subscriptions = [] ) {
+		$detached_messages = '';
+		foreach ( $subscriptions as $subscription ) {
+			$customer_payment_method_link = sprintf(
+				'<a href="%s">%s</a>',
+				esc_url( $subscription['change_payment_method_url'] ),
+				esc_html(
+				/* translators: this is a text for a link pointing to the customer's payment method page */
+					__( 'Payment method page &rarr;', 'woocommerce-gateway-stripe' )
+				)
+			);
+			$customer_stripe_page = sprintf(
+				'<a href="%s">%s</a>',
+				esc_url( self::STRIPE_CUSTOMER_PAGE_BASE_URL . $subscription['customer_id'] ),
+				esc_html(
+				/* translators: this is a text for a link pointing to the customer's page on Stripe */
+					__( 'Stripe customer page &rarr;', 'woocommerce-gateway-stripe' )
+				)
+			);
+			$detached_messages .= sprintf(
+			/* translators: %1$s is the subscription ID. %2$s is a customer payment method page. %3$s is the customer's page on Stripe */
+				__( '#%1$s: %2$s | %3$s<br/>', 'woocommerce-gateway-stripe' ),
+				$subscription['id'],
+				$customer_payment_method_link,
+				$customer_stripe_page
+			);
+		}
+		if ( ! empty( $detached_messages ) ) {
+			// phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped
+			$intro_message = sprintf(
+				/* translators: %s: subscriptions count */
+				_n(
+					'%s subscription is missing the payment method, <strong>preventing renewals</strong>. ',
+					'%s subscriptions are missing payment methods, <strong>preventing renewals</strong>. ',
+					count( $subscriptions ),
+					'woocommerce-gateway-stripe'
+				),
+				count( $subscriptions )
+			); /* phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped */
+			// phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped
+			$intro_message    .= __( 'To fix this, either: 1) Share the payment method page link with the customer to update it, or 2) Manually update the payment method in the subscription\'s billing details using a valid payment method from the customer\'s Stripe account. ', 'woocommerce-gateway-stripe' );
```

Comment:
Style question: is it worth using `"` quotes to avoid using `\'` in the string contents? The `\` characters make it harder to search for string matches.
```suggestion
			$intro_message    .= __( "To fix this, either: 1) Share the payment method page link with the customer to update it, or 2) Manually update the payment method in the subscription's billing details using a valid payment method from the customer's Stripe account. ", 'woocommerce-gateway-stripe' );
```

Reply:
Separately, might it be worth adding explicit `<br />` elements so the actions are super-clear?
```suggestion
			$intro_message    .= __( 'To fix this, either: <br />1) Share the payment method page link with the customer to update it, or <br />2) Manually update the payment method in the subscription\'s billing details using a valid payment method from the customer\'s Stripe account. ', 'woocommerce-gateway-stripe' );
```

Reply:
> Style question: is it worth using " quotes to avoid using \' in the string contents? The \ characters make it harder to search for string matches.

Previously, I remember we had always used single quotes for strings. I don't know if that changed here or if it was in another repository. But I have updated it manually now anyway in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4391/commits/8ae96ebffbc75e7ee0bab52110552a73a84bd14d (git does not allow me to apply changes to outdated files)

> Separately, might it be worth adding explicit <br /> elements so the actions are super-clear?

Makes sense! Added in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4391/commits/4ee7ff65f805d52a807c7ba593cbb343203c15e3